src/utils/vector-db/manager.ts

```diff
@@ -0,0 +1,213 @@
+              path: file.path,
+              mtime: file.stat.mtime,
+              content: chunk.pageContent,
+              metadata: {
+                startLine: chunk.metadata.loc.lines.from as number,
+                endLine: chunk.metadata.loc.lines.to as number,
+              },
+            }
+          })
+        }),
+      )
+    ).flat()
+
+    updateProgress?.({
+      completedChunks: 0,
+      totalChunks: contentChunks.length,
+      totalFiles: filesToIndex.length,
+    })
+
+    const embeddingProgress = { completed: 0 }
+    const embeddingChunks: InsertVector[] = []
+    const limit = pLimit(50)
+    const tasks = contentChunks.map((chunk) =>
+      limit(async () => {
+        const maxRetries = 5
+        const exponentialBackoff = 1.5
+        let retryCount = 0
+        let delay = 1000
+
+        // Retry wait times:
+        // 1. 1000 ms (1 second)
+        // 2. 1500 ms (1.5 seconds)
+        // 3. 2250 ms (2.25 seconds)
+        // 4. 3375 ms (3.375 seconds
+        // 5. 5062.5 ms (5.0625 seconds)
+        // This will be enough for OpenAI API rate limit (3000 RPM)
+
+        while (retryCount < maxRetries) {
+          try {
+            const embedding = await embeddingModel.getEmbedding(chunk.content)
+            const embeddedChunk = {
+              path: chunk.path,
+              mtime: chunk.mtime,
+              content: chunk.content,
+              embedding,
+              metadata: chunk.metadata,
+            }
+            embeddingChunks.push(embeddedChunk)
+            embeddingProgress.completed++
+            updateProgress?.({
+              completedChunks: embeddingProgress.completed,
+              totalChunks: contentChunks.length,
+              totalFiles: filesToIndex.length,
+            })
+            return
+          } catch (error) {
+            if (
+              error.status === 429 &&
+              error.message.toLowerCase().includes('rate limit')
+            ) {
+              // OpenAI API returns 429 status code for rate limit errors
+              // See: https://platform.openai.com/docs/guides/error-codes/api-errors
+              retryCount++
+              console.warn(
+                `Rate limit reached. Retrying in ${delay / 1000} seconds...`,
+              )
+              await new Promise((resolve) => setTimeout(resolve, delay))
+              delay *= exponentialBackoff
+            } else {
+              throw error
+            }
+          }
+        }
+        throw new Error(`Failed to process chunk after ${maxRetries} retries`)
+      }),
+    )
+
+    try {
+      await Promise.all(tasks)
+      await this.repository.insertVectors(embeddingChunks, embeddingModel)
+      await this.repository.save()
+    } catch (error) {
+      console.error('Error updating vault index:', error)
+      // Save processed chunks
+      await this.repository.insertVectors(embeddingChunks, embeddingModel)
+      await this.repository.save()
+      throw error
+    }
```

Comment:
Maybe it's better to add retry logic (using the library you mentioned) and just throw an error after the second retry?